library/psa_crypto.c

```diff
@@ -2302,39 +1786,34 @@ psa_status_t psa_import_key( const psa_key_attributes_t *attributes,
     if( status != PSA_SUCCESS )
         goto exit;
 
-#if defined(MBEDTLS_PSA_CRYPTO_SE_C)
-    if( driver != NULL )
+    /* In the case of a transparent key or an opaque key stored in local
+     * storage (thus not in the case of generating a key in a secure element
+     * or cryptoprocessor with storage), we have to allocate a buffer to
+     * hold the generated key material. */
+    if( slot->key.data == NULL )
     {
-        const psa_drv_se_t *drv = psa_get_se_driver_methods( driver );
-        /* The driver should set the number of key bits, however in
-         * case it doesn't, we initialize bits to an invalid value. */
-        size_t bits = PSA_MAX_KEY_BITS + 1;
-        if( drv->key_management == NULL ||
-            drv->key_management->p_import == NULL )
-        {
-            status = PSA_ERROR_NOT_SUPPORTED;
-            goto exit;
-        }
-        status = drv->key_management->p_import(
-            psa_get_se_driver_context( driver ),
-            slot->data.se.slot_number, attributes, data, data_length,
-            &bits );
+        status = psa_allocate_buffer_to_slot( slot, data_length );
         if( status != PSA_SUCCESS )
             goto exit;
-        if( bits > PSA_MAX_KEY_BITS )
-        {
-            status = PSA_ERROR_NOT_SUPPORTED;
-            goto exit;
-        }
-        slot->attr.bits = (psa_key_bits_t) bits;
     }
-    else
-#endif /* MBEDTLS_PSA_CRYPTO_SE_C */
+
+    bits = slot->attr.bits;
+    status = psa_driver_wrapper_import_key( attributes,
+                                            data, data_length,
+                                            slot->key.data,
+                                            slot->key.bytes,
+                                            &slot->key.bytes, &bits );
+    if( status != PSA_SUCCESS )
+        goto exit;
+
+    if( slot->attr.bits == 0 )
+        slot->attr.bits = (psa_key_bits_t) bits;
+    else if( bits != slot->attr.bits )
     {
-        status = psa_import_key_into_slot( slot, data, data_length );
-        if( status != PSA_SUCCESS )
-            goto exit;
+        status = PSA_ERROR_INVALID_ARGUMENT;
+        goto exit;
     }
+
     status = psa_validate_optional_attributes( slot, attributes );
```

Comment:
The call to `validate_optional_attributes` here will still rely on having a software implementation compiled in for RSA keys. That part of the validation would/should have been done as part of the `wrapper_import_key` call chain, such that the call tree after returning from a driver wrapper call does not rely on software libraries. Returning 'success' from the `wrapper_import_key` call evidently means the key is deemed valid.

Reply:
Yes, that makes sense to me. The RSA code checking the domain parameters in psa_validate_optional_attributes() should probably be moved to rsa_import_key() in psa_crypto_rsa.c. @gilles-peskine-arm, what do you thing about this?

Reply:
The validation of domain parameters also needs to be done when generating or deriving a key. Embedding it in import is definitely not right.

`psa_generate_key` should call `psa_validate_optional_attributes` (or maybe a refactored version of it). The reason it doesn't is that no use of domain parameters has been implemented yet in `psa_generate_key`, and the other validations are redundant for key generation, so it gets away with just requiring the domain parameters to be `NULL`. Which is a pity because the most asked feature that domain parameters would solve is generating an RSA key with e=3, and we're almost there.

Making it even harder for `psa_generate_key` to use domain parameters is not useful: if domain parameters are really blocking driver support, we should remove domain parameters. I hope we don't have to go that far.

Why can't we reserve domain parameters for the software case, and worry later about adding domain parameter support to the driver interface? If there's a driver, just require the domain parameters to be null.

----

The current state of domain parameters in Mbed TLS is unsatisfactory: there's some infrastructure, but not enough to do anything useful. Domain parameters were in a draft of the PSA Crypto API specification, but were removed from the 1.0 release because the experience of implementing it in Mbed TLS showed that the current form is a hassle. We haven't removed domain parameters from Mbed TLS because we do want the features they bring, but work on them is paused while we focus on other things such as driver support.

We've identified three uses for domain parameters in the currently declared algorithms:

* Specifying a non-default public exponent in RSA key generation. RSA is implemented in Mbed TLS, but a custom public exponent is not possible through the PSA API (it is possible through the classic API).
* Using a custom group for finite-field Diffie-Hellman. FFDH is implemented in Mbed TLS, but not yet through the PSA API, and even if it was a first step would be to only support predefined groups.
* Using DSA (the non-EC kind). I doubt that DSA will never be implemented in Mbed TLS.

Because the domain parameter interface is likely to change, I want to minimize its entanglement in the driver interface.

Reply:
From crypto_extra.h:
```
- For RSA keys (#PSA_KEY_TYPE_RSA_PUBLIC_KEY or #PSA_KEY_TYPE_RSA_KEY_PAIR), 
 *   the domain parameter data consists of the public exponent,                 
 *   represented as a big-endian integer with no leading zeros.                 
 *   This information is used when generating an RSA key pair.                  
 *   **When importing a key, the public exponent is read from the imported        
 *   key data and the exponent recorded in the attribute structure is ignored.**  
 *   As an exception, the public exponent 65537 is represented by an empty      
 *   byte string.  
 ```
Thus should we just remove the check of the RSA exponent when importing a key ? That sounds sensible to me.

Reply:
This text was added in https://github.com/ARMmbed/mbedtls-psa/pull/263 (private link). Looking back, it described the implementation at the time (import didn't check the RSA exponent), but I think this was an implementation limitation (domain parameters for RSA were a new thing) and not a desirable specification. Having a function ignore an explicit input is not a good design. https://github.com/ARMmbed/mbed-crypto/pull/101 added the check, which I think was the right thing, but forgot to update the documentation.

Reply:
Thanks for not blocking the PR on this. I agree that there is an issue and I've described it [here](https://github.com/ARMmbed/mbedtls/issues/4095). I am not addressing it here as I don't see a way of doing it in one or two not controversial commits.